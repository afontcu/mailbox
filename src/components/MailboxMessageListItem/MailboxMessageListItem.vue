<template>
  <a class="MailboxMessageListItem" 
     href="#"
     :class="[ classes ]"
     @click="selectMessage"
  >
    <div class="MailboxMessageListItem__header">
      <p class="MailboxMessageListItem__title">{{ sender }}</p>
      <button class="MailboxMessageListItem__remove-button"
              @click="handleRemove"
      >
        <icon-cross aria-label="Delete the message" />
      </button>
    </div>
    <p class="MailboxMessageListItem__subject">{{ subject }}</p>
    <p class="MailboxMessageListItem__date">{{ date }}</p>
  </a>
</template>

<script>
import { mapMutations, mapState } from 'vuex'
import IconCross from '@/components/IconCross'

export default {
  name: 'MailboxMessageListItem',
  components: {
    IconCross,
  },
  props: {
    uid: { type: String, required: true },
    sender: { type: String, required: true },
    subject: { type: String, required: true },
    date: { type: String, required: true },
  },
  computed: {
    ...mapState(['readMessages', 'selectedMessageId']),
    classes() {
      return {
        'MailboxMessageListItem--unread': !this.readMessages.includes(this.uid),
        'MailboxMessageListItem--selected': this.selectedMessageId === this.uid,
      }
    },
  },
  methods: {
    ...mapMutations(['markAsRead', 'markAsSelected', 'removeMessage']),
    selectMessage() {
      this.markAsSelected(this.uid)
      this.markAsRead(this.uid)
    },
    handleRemove(e) {
      e.stopPropagation()
      this.removeMessage(this.uid)
    },
  },
}
</script>

<style scoped>
:root {
  --item-border-width: 4px;
}

.MailboxMessageListItem + .MailboxMessageListItem {
  margin-top: var(--global-spacing);
}

.MailboxMessageListItem {
  padding: var(--global-spacing);
  position: relative;
  display: block;

  color: initial;
  text-decoration: none;
  border-left: var(--item-border-width) solid transparent;
  border-radius: var(--global-radius);
  background-color: #fff;
  box-shadow: 0 0 0 1px rgba(63, 63, 68, 0.05),
    0 1px 3px 0 rgba(63, 63, 68, 0.15);
  transition: all ease var(--global-transition);
}

.MailboxMessageListItem:not(.MailboxMessageListItem--selected):hover {
  background-color: var(--color-primary-light);
  box-shadow: 0px 2px 6px 0 rgba(63, 63, 68, 0.21);
}

.MailboxMessageListItem--unread {
  border-left: var(--item-border-width) solid var(--color-primary);
}

.MailboxMessageListItem--selected {
  transform: translateX(calc(2px + var(--global-spacing)));

  border-left: var(--item-border-width) solid var(--color-secondary);
  box-shadow: -5px 2px 6px 0 rgba(63, 63, 68, 0.21);
}

.MailboxMessageListItem__header {
  display: flex;
  justify-content: space-between;
}

.MailboxMessageListItem__remove-button {
  color: var(--color-text-muted);
  background: none;
  border: none;
  font: inherit;
  cursor: pointer;
}

.MailboxMessageListItem__title {
  font-weight: bold;
}

.MailboxMessageListItem__subject {
  margin: var(--global-spacing-small) 0 var(--global-spacing);
}

.MailboxMessageListItem__date {
  font-style: italic;
  color: var(--color-text-muted);
}
</style>
