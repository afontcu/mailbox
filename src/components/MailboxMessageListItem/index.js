import MailboxMessageListItem from './MailboxMessageListItem'

export default MailboxMessageListItem
