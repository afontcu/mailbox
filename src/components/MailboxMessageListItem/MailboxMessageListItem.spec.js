import MailboxMessageListItem from './index.js'
import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

const propsData = {
  uid: '1',
  sender: 'sender name',
  subject: 'subject',
  date: 'Tue 29 March, 12:52',
}

describe('MailboxMessageListItem.vue', () => {
  let store, mutations, state

  beforeEach(() => {
    state = {
      messages: [],
      readMessages: [],
    }
    mutations = {
      markAsRead: jest.fn(),
      markAsSelected: jest.fn(),
      removeMessage: jest.fn(),
    }
    store = new Vuex.Store({
      mutations,
      state,
    })
  })

  it('is a component', () => {
    const wrapper = mount(MailboxMessageListItem, {
      propsData,
      store,
      localVue,
    })
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('renders main node', () => {
    const wrapper = mount(MailboxMessageListItem, {
      propsData,
      store,
      localVue,
    })
    expect(wrapper.classes()).toContain('MailboxMessageListItem')
  })

  it('renders a snapshot', () => {
    const wrapper = mount(MailboxMessageListItem, {
      propsData,
      store,
      localVue,
    })
    expect(wrapper.html()).toMatchSnapshot('default')
  })
})
