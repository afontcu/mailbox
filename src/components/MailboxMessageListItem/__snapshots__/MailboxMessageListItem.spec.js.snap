// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailboxMessageListItem.vue renders a snapshot: default 1`] = `
<a href="#" class="MailboxMessageListItem MailboxMessageListItem--unread">
  <div class="MailboxMessageListItem__header">
    <p class="MailboxMessageListItem__title">sender name</p> <button class="MailboxMessageListItem__remove-button"><svg viewPort="0 0 12 12" width="12" height="12" aria-label="Delete the message" version="1.1" xmlns="http://www.w3.org/2000/svg">
        <line x1="1" y1="11" x2="11" y2="1" stroke="currentColor" stroke-width="2"></line>
        <line x1="1" y1="1" x2="11" y2="11" stroke="currentColor" stroke-width="2"></line>
      </svg></button>
  </div>
  <p class="MailboxMessageListItem__subject">subject</p>
  <p class="MailboxMessageListItem__date">Tue 29 March, 12:52</p>
</a>
`;
